File: ios/Crew/Storage/AchievementFacts.swift
// SPEC: README kind achievements — the SOLO counters derived from the phone's own store (posts, completed sessions, PRs) plus
// the engine's tallies, so an unlock can fold into the celebration offline (E8, E6). The crew counters (reactions given,
// crew full pulse) are the server's: they arrive with the reconciled state (5.6.3), never computed here. Twin of
// web/src/lib/achievement-facts.ts. WRITTEN — UNVERIFIED (needs Mac).

import Foundation
import SwiftData

@MainActor
enum AchievementFacts {
    static func soloCounters(for userId: String, state: GamificationState, store: Store) throws -> AchievementCounters {
        let completed = try store.context.fetch(FetchDescriptor<LocalSession>(predicate: #Predicate { $0.userId == userId && $0.status == "completed" }))
        var counters = AchievementCounters()
        counters.postsTotal = try store.allPosts(for: userId).count
        counters.workoutsCompleted = completed.count
        counters.currentStreak = state.currentStreak
        counters.perfectWeeks = state.tallies.perfectWeeks
        counters.shieldsConsumed = state.tallies.shieldsConsumed
        counters.comebacks = state.tallies.comebacks
        counters.prCount = PersonalRecords.prCount(completed.map(recordSession), accountUnit: AuthStore.shared.weightUnit)
        counters.crewJoined = try store.crewSnapshot() == nil ? 0 : 1
        return counters
    }

    static func recordSession(_ session: LocalSession) -> RecordSession {
        RecordSession(completedAt: session.completedAt ?? session.startedAt, exercises: session.exercises.map { exercise in
            // A9: each set carries the unit it was ENTERED in, so bests compare on one normalised scale
            RecordExercise(exerciseId: exercise.exerciseId, name: exercise.name, sets: exercise.sets.map { RecordSet(done: $0.done, isWarmup: $0.isWarmup, weight: $0.weight, weightUnit: $0.weightUnit) })
        })
    }

    // Flow 3 PR celebration: this session against every earlier completed one (the same rule the server counts)
    static func newRecords(in session: LocalSession, store: Store) throws -> [String] {
        let clientId = session.clientId
        let userId = session.userId
        let earlier = try store.context.fetch(FetchDescriptor<LocalSession>(predicate: #Predicate { $0.userId == userId && $0.status == "completed" && $0.clientId != clientId }))
        return PersonalRecords.newRecords(recordSession(session).exercises, earlier: earlier.map(recordSession), accountUnit: AuthStore.shared.weightUnit)
    }
}

File: web/src/lib/achievement-facts.ts
// SPEC: README kind achievements — the counters are FACTS derived from the user's documents plus the engine's tallies;
// the pass (achievementsEarned) runs after every recompute (5.6.4) and earned ids are only ever added (V35).
// Server twin of ios/Crew/Storage/AchievementFacts.swift (which derives the solo counters from the local store).
import type { ObjectId } from "mongodb";
import { crewMemberships, posts, reactions, sessions } from "@/lib/db";
import type { AchievementCounters } from "@/lib/engine/achievements";
import { fullPulseDays, fullPulseWeeks, type MemberFacts } from "@/lib/engine/crew-rules";
import type { GamificationState } from "@/lib/engine/gamification";
import { prCount, type RecordSession } from "@/lib/engine/personal-records";
import type { CrewMembershipDoc } from "@/lib/documents-social";

export async function achievementCounters(userId: ObjectId, state: GamificationState, todayKey: string, accountUnit: "lb" | "kg" = "lb"): Promise<AchievementCounters> {
  const [postsTotal, completed, reactionsGiven, membership] = await Promise.all([
    (await posts()).countDocuments({ userId, deletedAt: null }),
    (await sessions()).find({ userId, status: "completed" }, { projection: { completedAt: 1, exercises: 1 } }).toArray(),
    (await reactions()).countDocuments({ userId }),
    (await crewMemberships()).findOne({ userId }),
  ]);
  const history: RecordSession[] = completed.map((doc) => ({ completedAt: (doc.completedAt ?? new Date(0)).toISOString(), exercises: doc.exercises.map((row) => ({ exerciseId: row.exerciseId, name: row.name, sets: row.sets })) }));
  const crew = membership === null ? { days: 0, weeks: 0 } : await crewFullPulse(membership, todayKey);
  return {
    postsTotal, workoutsCompleted: completed.length, currentStreak: state.currentStreak, perfectWeeks: state.tallies.perfectWeeks, prCount: prCount(history, accountUnit), // A9: compare on one normalised scale
    shieldsConsumed: state.tallies.shieldsConsumed, comebacks: state.tallies.comebacks, crewJoined: membership === null ? 0 : 1, reactionsGiven,
    crewFullPulseDays: crew.days, crewFullPulseWeeks: crew.weeks,
  };
}

// Membership as of each day (V40) from the memberships that exist now; every member's posts since this user joined
async function crewFullPulse(membership: CrewMembershipDoc, todayKey: string): Promise<{ days: number; weeks: number }> {
  const memberDocs = await (await crewMemberships()).find({ crewId: membership.crewId }).toArray();
  const members: MemberFacts[] = memberDocs.map((doc) => ({ userId: doc.userId.toHexString(), joinedDayKey: doc.joinedDayKey }));
  const fromDay = membership.joinedDayKey;
  const memberPosts = await (await posts()).find({ userId: { $in: memberDocs.map((doc) => doc.userId) }, deletedAt: null, dayKey: { $gte: fromDay, $lte: todayKey } }, { projection: { userId: 1, dayKey: 1 } }).toArray();
  const postFacts = memberPosts.map((doc) => ({ userId: doc.userId.toHexString(), dayKey: doc.dayKey }));
  return { days: fullPulseDays(members, postFacts, fromDay, todayKey), weeks: fullPulseWeeks(members, postFacts, fromDay, todayKey) };
}
